src/libopensc/card-eoi.c

```diff
@@ -0,0 +1,496 @@
+	if (!(env->algorithm_flags & ALREADY_PROCESSED)) {
+		privdata->key_len = (env->algorithm_ref + 7)/8;
+		privdata->sec_env = (struct sc_security_env *)env;
+		privdata->se_num = se_num;
+		LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+	}
+
+	if (env->operation != SC_SEC_OPERATION_SIGN)
+		return SC_ERROR_NOT_SUPPORTED;
+	if (!(env->flags & SC_SEC_ENV_KEY_REF_PRESENT))
+		return SC_ERROR_INVALID_ARGUMENTS;
+	if (env->key_ref_len != 1)
+		return SC_ERROR_INVALID_ARGUMENTS;
+	if (env->algorithm != SC_ALGORITHM_EC)
+		return SC_ERROR_INVALID_ARGUMENTS;
+	sc_format_apdu(card, &apdu, SC_APDU_CASE_3_SHORT, 0x22, 0x81, 0xB6);
+	sbuf[0] = 0x91;
+	sbuf[1] = 0x02;
+	switch (env->algorithm_flags & ~ALREADY_PROCESSED) {
+		case SC_ALGORITHM_ECDSA_HASH_SHA1:
+			sbuf[2] = 0x11;
+			break;
+		case SC_ALGORITHM_ECDSA_HASH_SHA256:
+			sbuf[2] = 0x21;
+			break;
+		case SC_ALGORITHM_ECDSA_RAW | SC_ALGORITHM_ECDSA_HASH_NONE:
+			sbuf[2] = 0x22;
+			break;
+		default:
+			return SC_ERROR_INVALID_ARGUMENTS;
+	}
+	for (i = 0; i < MAX_OBJECTS && privdata->prkey_mappings[i][1]; i++) {
+		if (privdata->prkey_mappings[i][0] == env->key_ref[0])
+			break;
+	}
+	if (i == MAX_OBJECTS)
+		return SC_ERROR_INVALID_ARGUMENTS;
+	sbuf[3] = privdata->prkey_mappings[i][1];
+	apdu.lc = 4;
+	apdu.datalen = 4;
+	apdu.data = sbuf;
+	if (se_num > 0) {
+		r = sc_lock(card);
+		LOG_TEST_RET(card->ctx, r, "sc_lock() failed");
+		locked = 1;
+	}
+	if (apdu.datalen) {
+		r = sc_transmit_apdu(card, &apdu);
+		if (r) {
+			sc_log(card->ctx, "%s: APDU transmit failed", sc_strerror(r));
+			goto err;
+		}
+		r = sc_check_sw(card, apdu.sw1, apdu.sw2);
+		if (r) {
+			sc_log(card->ctx, "%s: Card returned error", sc_strerror(r));
+			goto err;
+		}
+	}
+	if (se_num <= 0)
+	    return 0;
+	sc_unlock(card);
+	LOG_TEST_RET(card->ctx, r, "APDU transmit failed");
+
+	return sc_check_sw(card, apdu.sw1, apdu.sw2);
+err:
+	if (locked)
+		sc_unlock(card);
+	return r;
+
+	LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+}
+
+static int eoi_compute_signature(struct sc_card *card, const u8 * data, size_t data_len, u8 *out, size_t outlen)
+{
+	struct eoi_privdata *privdata = (struct eoi_privdata *)card->drv_data;
+
+	LOG_FUNC_CALLED(card->ctx);
+
+	if (!privdata)
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_INTERNAL);
+
+	/*
+	 * Guess the correct mode. If the size is less than the full-key-len, it must be a hash then
+	 */
+	if (privdata->key_len != data_len) {
+		switch (data_len) {
+			case SHA_DIGEST_LENGTH:
+				privdata->sec_env->algorithm_flags = SC_ALGORITHM_ECDSA_HASH_SHA1;
+				break;
+			case SHA256_DIGEST_LENGTH:
+				privdata->sec_env->algorithm_flags = SC_ALGORITHM_ECDSA_HASH_SHA256;
+				break;
+		}
+	}
+	/* Now we know which hash is used */
+	privdata->sec_env->algorithm_flags |= ALREADY_PROCESSED;
+
+	/* Perform the true set_security_env */
+	LOG_TEST_RET(card->ctx, eoi_set_security_env(card, privdata->sec_env, privdata->se_num),
+			"set_security_env failed");
```

Comment:
on the first sight, this looks suspicious. Why this can not be handled directly in the set_security_env? The rest of the opensc should make sure correct flags are passed already at that time.

Reply:
Here's how the eOI card is doing the signing (from what I gather from analyzing the APDU trace/log of the 'official' PKCS#11):
1) hashing of the data is done off-card, in the PKCS#11 library;
2) nonetheles type of the hash used is passed as a parameter in a 'set security environment' APDU.

It is my undestanding, that for OpenSC to do the off-card hashing, `SC_ALGORITHM_ECDSA_RAW | SC_ALGORITHM_ECDSA_HASH_NONE` (and not `SC_ALGORITHM_ECDSA_HASH_SHAx`) has to be specified when registering an algorithm using `_sc_card_add_ec_alg()`. But this also means that `SC_ALGORITHM_ECDSA_RAW | SC_ALGORITHM_ECDSA_HASH_NONE` is passed as part of the `struct sc_security_env` to the `set_security_env()` and so the true type of hash used is not known at this time and has to be deduced from the size of the data-to-be-signed in the `compute_signature()`. It would be nice, if the hash mechanism specified in a` C_SignInit` was somehow part of the `struct sc_security_env`, but it isn't.
So I don't see how it can be done differently?

Reply:
Yes, normally this is all what is needed. But unfortunately not in this case.

eOI card  `set_sec_env()` APDU has (at least 3 forms):

1) The data to be signed is a 20 byte (generated off-card) SHA1 hash: SET SECURITY APDU is `00 22 81 B6 ... 11 [keyref]`
2) The data to be signed is a 32 byte (generated off-card)  SHA256 hash: SET SECURITY APDU is `00 22 81 B6 ... 21 [keyref]`
3) The data to be signed is raw key-len : SET SECURITY APDU is `00 ... 81 B6 ... 22 [keyref]`

And there is no way to know which form to use if `env->algorithm_flags == SC_ALGORITHM_ECDSA_RAW` and hashing is done off-card?

I could try to always use the "raw" SET SECURITY APDU, but that's not how the "official" PKCS#11 lib does it.

Reply:
Well, I didn't find a better solution. But, of course, I am open to suggestions...